pkg/utils/k8s/fake.go

```diff
@@ -20,56 +20,103 @@ func NewFakeClient(initObjs ...runtime.Object) Client {
 }
 
 var (
-	_ Client              = failingClient{}
+	_ Client              = &FailingClient{}
 	_ client.StatusWriter = failingStatusWriter{}
 )
 
-type failingClient struct {
-	err error
+// FailingClient is a k8s client that fails with a given error by default.  When erroring
+// is false by calling 'DisableFailing', then underlying client is called.
+type FailingClient struct {
+	client   Client
+	erroring bool
+	err      error
+}
+
+// NewFailingClient returns a client that returns the provided error when called and 'DisableFailing' has not been called.
+// After 'DisableFailing' is called, the client calls the underlying k8s client.
+func NewFailingClient(err error, initObjs ...runtime.Object) Client {
+	return &FailingClient{
+		client:   fake.NewClientBuilder().WithRuntimeObjects(initObjs...).WithScheme(clientgoscheme.Scheme).Build(),
+		erroring: true,
+		err:      err,
+	}
+}
+
+// Get satisfies the controller-runtime client interface.
+func (fc *FailingClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	if !fc.erroring {
+		return fc.client.Get(ctx, key, obj)
+	}
+	return fc.err
 }
 
-// NewFailingClient returns a client that always returns the provided error when called.
-func NewFailingClient(err error) Client {
-	return failingClient{err: err}
+// List satisfies the controller-runtime client interface.
+func (fc *FailingClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	if !fc.erroring {
```

Comment:
I think I am slightly opposed to changing the failing client into a sometimes failing client like this. See my comment to the one test case where you are using it.

I think the argument would be to keep the logic in these shared test fixtures minimal.